File: datasets/C++/Mistral/DataStructures/hard4.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

using namespace std;

int main() {
    int N;
    cout << "Enter the number of elements in the array: ";
    cin >> N;
    vector<int> A(N);
    cout << "Enter the elements of the array separated by space: ";
    for (int i = 0; i < N; ++i) {
        cin >> A[i];
    }

    int max_si_value = INT_MIN;

    for (int L = 0; L < N; ++L) {
        for (int R = L + 1; R < N; ++R) {
            vector<int> subarray(A.begin() + L, A.begin() + R + 1);
            sort(subarray.begin(), subarray.end());
            int M1 = subarray[0];
            int M2 = subarray[1];
            int Si = (((M1 & M2) ^ (M1 | M2)) & (M1 ^ M2));
            if (Si > max_si_value) {
                max_si_value = Si;
            }
        }
    }

    cout << "The maximum possible value of Si is: " << max_si_value << endl;

    return 0;
}

File: datasets/Python/Mistral/DataStructures/hard4.py
def max_si():
    N = int(input("Enter the number of elements in the array: "))
    A = list(map(int, input("Enter the elements of the array separated by space: ").split()))

    max_si_value = 0

    for L in range(N):
        for R in range(L + 1, N):
            subarray = A[L:R+1]
            subarray_sorted = sorted(subarray)
            M1 = subarray_sorted[0]
            M2 = subarray_sorted[1]
            Si = (((M1 & M2) ^ (M1 | M2)) & (M1 ^ M2))
            if Si > max_si_value:
                max_si_value = Si

    print("The maximum possible value of Si is:", max_si_value)

max_si()